plugins/BEdita/Core/src/Model/Action/SignupUserAction.php

```diff
@@ -181,13 +205,53 @@ protected function validate(array $data)
                 'provider' => 'bedita',
             ]);
 
+        if (!empty($this->getConfig('roles'))) {
+            $validator->requirePresence('roles');
+        }
+
+        $validator->add('roles', 'validateRoles', [
+            'rule' => [$this, 'validateRoles'],
+        ]);
+
         return $validator->errors($data);
     }
 
+    /**
+     * Validate roles against allowed signup roles configured.
+     * In addtion roles can't contain ADMIN_ROLE.
+     *
+     * @param string|array $roles The roles to check
+     * @return true|string
+     */
+    public function validateRoles($roles)
+    {
+        $allowedRoles = (array)$this->getConfig('roles');
+        if (empty($allowedRoles) && !empty($roles)) {
+            return __d('bedita', 'Roles are not allowed on signup');
+        }
+
+        $adminRoleName = $this->Roles->get(RolesTable::ADMIN_ROLE)->get('name');
+
+        $roles = (array)$roles;
+        $message = '{0} not allowed on signup';
```

Comment:
Maybe `'Role {0} not allowed on signup'` ?

Reply:
I have simplified because the error shown in the exception will contain in the detail

```php
[
    'roles' => [
        'validateRoles' => 'third_role, fourth_role not allowed on signup',
    ],
]
```

that contains already information about "roles".

Moreover I wanted avoid to distinguish between singular and plural :smile_cat: